File: apple/Sources/Vifu/VifuGatewayPairingCode.swift
import CryptoKit
import Foundation

public enum VifuGatewayPairingCodeError: LocalizedError {
    case invalidCode

    public var errorDescription: String? {
        "This Vifu pairing code is invalid or has expired."
    }
}

/// A one-time Agent Gateway enrollment bound to one Vifu Server URL.
public struct VifuGatewayPairingCode: Sendable {
    public let serverURL: String
    public let enrollmentToken: String
    public let serverCertificateDER: Data?
    public let serverCertificateSHA256: String?

    public init(code: String) throws {
        guard let components = URLComponents(string: code),
              components.scheme == "vifu",
              components.host == "gateway",
              components.path == "/enroll",
              let serverURL = components.queryItems?.first(
                where: { $0.name == "server" }
              )?.value,
              Self.isHTTPSOrigin(serverURL),
              let token = components.queryItems?.first(
                where: { $0.name == "token" }
              )?.value,
              token.hasPrefix("vifu_ge_"),
              token.utf8.count == 72,
              token.utf8.dropFirst(8).allSatisfy(Self.isASCIIHexDigit)
        else {
            throw VifuGatewayPairingCodeError.invalidCode
        }

        let encodedCertificate = components.queryItems?.first(
            where: { $0.name == "certificate" }
        )?.value
        let fingerprint = components.queryItems?.first(
            where: { $0.name == "fingerprint" }
        )?.value
        let certificate: Data?
        switch (encodedCertificate, fingerprint) {
        case (nil, nil):
            certificate = nil
        case let (encoded?, fingerprint?):
            guard let decoded = Data(base64Encoded: encoded),
                  !decoded.isEmpty,
                  fingerprint == Self.fingerprint(decoded)
            else {
                throw VifuGatewayPairingCodeError.invalidCode
            }
            certificate = decoded
        default:
            throw VifuGatewayPairingCodeError.invalidCode
        }

        self.serverURL = serverURL
        enrollmentToken = token
        serverCertificateDER = certificate
        serverCertificateSHA256 = fingerprint
    }

    private static func isHTTPSOrigin(_ value: String) -> Bool {
        guard let components = URLComponents(string: value),
              components.scheme?.lowercased() == "https",
              let host = components.host,
              !host.isEmpty,
              components.user == nil,
              components.password == nil,
              components.query == nil,
              components.fragment == nil
        else {
            return false
        }
        return components.path.isEmpty || components.path == "/"
    }

    private static func fingerprint(_ certificateDER: Data) -> String {
        let digest = SHA256.hash(data: certificateDER)
        return "sha256:" + digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func isASCIIHexDigit(_ byte: UInt8) -> Bool {
        (48 ... 57).contains(byte) || (65 ... 70).contains(byte) || (97 ... 102).contains(byte)
    }
}

File: integrations/android/core-shared/src/main/kotlin/dev/vifu/android/VifuGatewayPairingCode.kt
package dev.vifu.android

import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.Base64

/** A one-time Agent Gateway enrollment bound to one Vifu Server URL. */
class VifuGatewayPairingCode(code: String) {
    val serverUrl: String
    val enrollmentToken: String
    private val certificateDer: ByteArray?
    val serverCertificateDer: ByteArray?
        get() = certificateDer?.copyOf()
    val serverCertificateSha256: String?

    init {
        val uri = parseUri(code)
        require(uri.scheme == "vifu" && uri.host == "gateway" && uri.path == "/enroll") {
            INVALID_CODE
        }
        val query = parseQuery(requireNotNull(uri.rawQuery) { INVALID_CODE })
        val server = requireNotNull(query["server"]) { INVALID_CODE }
        require(isHttpsOrigin(server)) { INVALID_CODE }
        val token = requireNotNull(query["token"]) { INVALID_CODE }
        require(ENROLLMENT_TOKEN.matches(token)) { INVALID_CODE }

        val encodedCertificate = query["certificate"]
        val fingerprint = query["fingerprint"]
        val certificate = when {
            encodedCertificate == null && fingerprint == null -> null
            encodedCertificate != null && fingerprint != null -> {
                val decoded = runCatching { Base64.getDecoder().decode(encodedCertificate) }
                    .getOrNull()
                require(decoded != null && decoded.isNotEmpty() && fingerprint == fingerprint(decoded)) {
                    INVALID_CODE
                }
                decoded
            }
            else -> throw IllegalArgumentException(INVALID_CODE)
        }

        serverUrl = server
        enrollmentToken = token
        certificateDer = certificate
        serverCertificateSha256 = fingerprint
    }

    fun connectionConfig(captureTraceContent: Boolean = false) = VifuConnectionConfig(
        serverUrl = serverUrl,
        serverCertificateDer = certificateDer?.copyOf(),
        captureTraceContent = captureTraceContent,
        enrollmentToken = enrollmentToken,
    )

    private companion object {
        const val INVALID_CODE = "This Vifu pairing code is invalid or has expired."
        val ENROLLMENT_TOKEN = Regex("^vifu_ge_[0-9a-fA-F]{64}$")

        fun parseUri(code: String): URI = runCatching { URI(code.trim()) }
            .getOrElse { throw IllegalArgumentException(INVALID_CODE, it) }

        fun parseQuery(rawQuery: String): Map<String, String> = rawQuery
            .split('&')
            .map { item ->
                val parts = item.split('=', limit = 2)
                require(parts.size == 2) { INVALID_CODE }
                decode(parts[0]) to decode(parts[1])
            }
            .also { pairs ->
                require(pairs.map { it.first }.distinct().size == pairs.size) { INVALID_CODE }
            }
            .toMap()

        fun decode(value: String): String = runCatching {
            URLDecoder.decode(value, StandardCharsets.UTF_8)
        }.getOrElse { throw IllegalArgumentException(INVALID_CODE, it) }

        fun isHttpsOrigin(value: String): Boolean {
            val uri = runCatching { URI(value) }.getOrNull() ?: return false
            return uri.scheme?.lowercase() == "https" &&
                !uri.host.isNullOrEmpty() &&
                uri.userInfo == null &&
                uri.query == null &&
                uri.fragment == null &&
                (uri.path.isNullOrEmpty() || uri.path == "/")
        }

        fun fingerprint(certificate: ByteArray): String = MessageDigest.getInstance("SHA-256")
            .digest(certificate)
            .joinToString(prefix = "sha256:", separator = "") { "%02x".format(it) }
    }
}
